lib/src/third_party/diff_match_patch/utils.dart

```diff
@@ -0,0 +1,253 @@
+part of 'diff.dart';
+
+/// Determine the common prefix of two strings.
+///
+/// [text1] is the first string. [text2] is the second string.
+/// Returns the number of characters common to the start of each string.
+@visibleForTesting
+int diffCommonPrefix(String text1, String text2) {
+  // TODO: Once Dart's performance stabilizes, determine if linear or binary
+  // search is better.
+  // Performance analysis: https://neil.fraser.name/news/2007/10/09/
+  final n = min(text1.length, text2.length);
+
+  for (var i = 0; i < n; i++) {
+    if (text1[i] != text2[i]) {
+      return i;
+    }
+  }
+  return n;
+}
+
+/// Determine the common suffix of two strings.
+///
+/// [text1] is the first string. [text2] is the second string.
+/// Returns the number of characters common to the end of each string.
+@visibleForTesting
+int diffCommonSuffix(String text1, String text2) {
+  // TODO: Once Dart's performance stabilizes, determine if linear or binary
+  // search is better.
+  // Performance analysis: https://neil.fraser.name/news/2007/10/09/
+  final text1Length = text1.length;
+  final text2Length = text2.length;
+  final n = min(text1Length, text2Length);
+
+  for (var i = 1; i <= n; i++) {
+    if (text1[text1Length - i] != text2[text2Length - i]) {
+      return i - 1;
+    }
+  }
+  return n;
+}
+
+/// Split a text into a list of strings. Reduce the texts to a string of
+/// hashes where each Unicode character represents one line.
+///
+/// [text] is the string to encode.
+/// [lineArray] is a List of unique strings.
+/// [lineHash] is a Map of strings to indices.
+/// [maxLines] is the maximum length for lineArray.
+/// Returns an encoded string.
+String _diffLinesToCharsMunge(
```

Comment:
I don't understand why we want to do this. Is this from a reference implementation, then please provide a link.